Shrink SignalR JS client (to 11kB compressed)

benaadams opened the pull request:
### Highlights

Significantly shrinks the uncompressed sizes of SignalR, MsgPack and Blazor.Server; and their corresponding compressed sizes. Leading to faster downloads and less JS parsing and compiling by the browser, allowing for faster start-up.

**Shrinks `signalr.min.js` by `91 kB`** (now `10 kB` br compressed, `12 kB` gz compressed)
**Shrinks `blazor.server.js` by `96 kB`** (now `28 kB` br compressed, `33 kB` gz compressed)
**Shrinks add-on msgpack by `112 kB`** (now `8 kB` br compressed, `9 kB` gz compressed)
**Don't need to add an additional `script` reference to `msgpack5.js` when using mspack**

### Main changes

* Updates TypeScript and dependencies to latest versions
* uglify-js does not support ES6+; so switched to terser which is also webpack's default as it supports new JS language features (like `class`)
* Marked the SingalR modules as `"sideEffects": false`; so tree-shaking can work better
* Dropped the `"es6-promise/dist/es6-promise.auto.js"` pollyfill
* Changed TS to output `es2019` rather than `es5` and dropped the "es2015.promise" and "es2015.iterable" pollyfills
  Note:  es2019 is already being used by Blazor
https://github.com/dotnet/aspnetcore/blob/a28da74c363454b78fbcf042e1a94dfc6e255e1a/src/Components/Web.JS/tsconfig.json#L8-L9
* Moved to `msgpack5` from `@msgpack/msgpack` as it requires less pollyfills, is typescript & module aware

### Required docs change

* Now only need the `@microsoft/signalr-protocol-msgpack` package for msgpack; don't need to include `msgpack5` which means its an additional `29 kB` rather than `140 kB` to use msgpack rather than Json (protocol-msgpack `16 kb` + msgpack5 `124 kB`)

### Sizes

SignalR **before** (130 kB minimized)
SignalR **after** (39 kB minimized, 10 kB br compressed, 12 kB gz compressed)

![image](https://user-images.githubusercontent.com/1142958/108805203-70540c80-7597-11eb-97ce-fd37a6a920d8.png)


Shrinks `blazor.server.js` by 45% from `212 kb` to `116 kB` (to `28 kB` br compressed, `33 kB` gz compressed)

![image](https://user-images.githubusercontent.com/1142958/108664702-33224880-74cb-11eb-88ae-68965e7f5b91.png)


Addresses https://github.com/dotnet/aspnetcore/issues/30319
Addresses https://github.com/dotnet/aspnetcore/issues/29904

benaadams committed:
Shrink SignalR JS client

BrennanConroy commented:
Nice, is the main benefit from using terser instead of uglify-js?

We'll need to discuss some of the version updates with the team before merging this.

benaadams commented:
> is the main benefit from using terser instead of uglify-js

* uglify-js does not support ES6+; so switched to terser which is also webpack's default as it supports new JS language features (like `class`)

* Marked the SingalR modules as `"sideEffects": false`; so tree-shaking can work better

* Dropped the `"es6-promise/dist/es6-promise.auto.js"` pollyfill

* Changed TS to output `es2019` rather than `es5` and dropped the "es2015.promise" and "es2015.iterable" pollyfills

benaadams commented:
`ERROR: 8:22  array-type  Array type using 'Array<T>' is forbidden. Use 'T[]' instead.`

😅 I was getting the opposite complaint when compiling locally

benaadams commented:
Note:  es2019 is already being used by Blazor

https://github.com/dotnet/aspnetcore/blob/a28da74c363454b78fbcf042e1a94dfc6e255e1a/src/Components/Web.JS/tsconfig.json#L8-L9

javiercn commented on src/SignalR/clients/ts/FunctionalTests/package.json:
```diff
@@ -30,9 +30,11 @@
     "rxjs": "^6.3.3",
     "ts-node": "^8.6.2",
     "typescript": "^3.7.5",
+    "webpack": "^5.23.0",
```
This is nice, didn't you get into issues with common pollyfills not being included anymore?

I did started down this path but somehow I produced a bigger bundle for Blazor than the one we currently produce.

benaadams replied:
Yes; it was just `process` in the tests, though is more problematic with the blazor js (as that's missing more pollyfills)

javiercn commented:
> Note: es2019 is already being used by Blazor

Yep, I'm not sure if SignalR needs to support IE11 or so though. That said, it shouldn't be hard to produce a separate ES5 bundle I think.

I've put a link on this PR to the issue we have for doing this on Blazor so that we don't loose track of it.

Thanks @benaadams for doing this work!

benaadams commented:
> Yep, I'm not sure if SignalR needs to support IE11 or so though.

6.0 is coming out in Nov 2021; Teams has already dropped support for IE11 (Aug 17 2020) and Microsoft 365 drops support on Aug 17 2021 and Microsoft Edge supports Windows 7 so...

davidfowl commented:
We don't support IE11. You need to add polyfills to get IE11 support AFAIK

benaadams commented:
Can go smaller :)

![image](https://user-images.githubusercontent.com/1142958/108602922-cc9f0c80-739c-11eb-85a9-35336fa4a3e3.png)

benaadams commented:
protocol-msgpack `16kb` -> `7kB` (2kB compressed)

![image](https://user-images.githubusercontent.com/1142958/108603129-091f3800-739e-11eb-8914-966853eaa4c8.png)

javiercn commented:
This is great!

benaadams committed:
Update Blazor Js to Webpack5

benaadams committed:
Move msgpack5 -> @msgpack/msgpack

benaadams commented:
Moved to `msgpack5` from `@msgpack/msgpack` as it requires less pollyfills, is typescript & module aware, and its from the msgpack team

Shrinks `blazor.server.js` by 45% from `212kb` to `116kB` (to `28kB` br compressed, `33kB` gz compressed)

![image](https://user-images.githubusercontent.com/1142958/108664702-33224880-74cb-11eb-88ae-68965e7f5b91.png)

benaadams commented:
**Required docs change**
* Now only need the `@microsoft/signalr-protocol-msgpack` package for msgpack; don't need to include `msgpack5` which means its an additional `29kB` rather than `140kB` to use msgpack rather than Json (protocol-msgpack `16kb` + msgpack5 `124kB`)

javiercn commented:
@benaadams really nice to see this coming along.

Let me know when you are ready for me to review, but so far what I've seen looks great.

SteveSandersonMS commented on src/Components/Web.JS/dist/Release/blazor.server.js:
```diff
@@ -1,25 +1 @@
-!function(e){var t={};function n(r){if(t[r])return t[r].exports;var o=t[r]={i:r,l:!1,exports:{}};return e[r].call(o.exports,o,o.exports,n),o.l=!0,o.exports}n.m=e,n.c=t,n.d=function(e,t,r){n.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:r})},n.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"_ ...
```
Unfortunately we do have to include this output file for `blazor.server.js` in source control. It's annoying that we have to, and frequently forces us to resolve merge conflicts. But it's needed due to our source build requirements for certain Linux distributions.

benaadams replied:
File is still there; is just a single line as doesn't need the node.js buffer module pollyfill for the browser

SteveSandersonMS commented on src/Components/Web.JS/package.json:
```diff
@@ -4,6 +4,7 @@
   "version": "0.0.1",
   "description": "",
   "main": "index.js",
+  "sideEffects": false,
```
If I understand [the meaning of this property](https://webpack.js.org/guides/tree-shaking/#clarifying-tree-shaking-and-sideeffects) correctly, it's not really accurate to say `sideEffects: false` since this isn't a pure module. It does have side effects, such as triggering the loading of Blazor apps, and the placement of APIs in global scope (e.g., `Blazor.*` and Emscripten's `Module.*`).

If this flag causes a meaningful reduction in size while not breaking functionality, then I'd be fine with lying to Webpack here, since we should be able to notice if anything new becomes broken in the future.

SteveSandersonMS replied:
Oh, I'm a bit confused now because https://github.com/dotnet/aspnetcore/pull/30320/files#diff-d40149c3a278492570fd82822aa94771c9fe60671c33a9508fa3576f0d9aee02R24 says `sideEffects: true`. Is that the real source of truth? If so, can this line be removed from `package.json`, or does it serve a different purpose?

benaadams replied:
`sideEffects` is more action at a distance; if one method can effect another method without any obvious coupling, which is one of the reasons I moved to an imported `Blazor` object rather than disparate method assignments on `window["Blazor"]` which are then called in uncoupled methods. i.e. "must be included" code, regardless of whether its called.

Might be getting away with it as the side effect methods are in the main startup path; however just in case added 
```ts
"sideEffects": [
  "src/GlobalExports.ts",
  "src/Platform/Mono/MonoPlatform.ts"
],
```

No change on overall size.

>  says `sideEffects: true` ...

That one is `optimization` object which says to use side effect processing in tree shaking

SteveSandersonMS commented on src/Components/Web.JS/src/Boot.Server.ts:
```diff
@@ -27,11 +27,11 @@ async function boot(userOptions?: Partial<CircuitStartOptions>): Promise<void> {
-  window['Blazor'].defaultReconnectionHandler = new DefaultReconnectionHandler(logger);
-  window['Blazor']._internal.InputFile = InputFile;
+  Blazor.defaultReconnectionHandler = new DefaultReconnectionHandler(logger);
+  Blazor._internal.InputFile = InputFile;
```
Thanks for sorting out some of this syntax weirdness! We've had an issue on the backlog for a while to come and tidy up around here in various ways, including this one.

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Circuits/DefaultReconnectDisplay.ts:
```diff
@@ -52,7 +53,7 @@ export class DefaultReconnectDisplay implements ReconnectDisplay {
         // - false to mean we reached the server, but it rejected the connection (e.g., unknown circuit ID)
         // - exception to mean we didn't reach the server (this can be sync or async)
-        const successful = await window['Blazor'].reconnect();
+        const successful = await (Blazor?.reconnect as any)();
```
Sorry if I'm missing something obvious, but why is it `Blazor?.reconnect` here and not just `Blazor.reconnect`? AFAICT `Blazor` would always be populated here.

SteveSandersonMS replied:
Perhaps it's a typo. In `DefaultReconnectionHandler.ts`, there's no `?` on the equivalent.

benaadams replied:
Also has it in `DefaultReconnectionHandler.ts`
```ts
this._reconnectCallback = reconnectCallback || (() => (Blazor.reconnect as any)());
```

Its because `Blazor` is created in `GlobalExports.ts` however `.reconnect` is assigned in `Boot.Server.ts` or `Boot.WebAssembly.ts` so it has to be nullable (due to late assignment); TypeScript then gives error "can't invoke method that may be undefined"; so need the `as any` cast to get around that compile error.

SteveSandersonMS commented on src/Components/Web.JS/src/Services/ComponentDescriptorDiscovery.ts:
```diff
@@ -133,8 +133,7 @@ function getComponentComment(commentNodeIterator: ComponentCommentIterator, type
   if (candidateStart.textContent) {
-    const componentStartComment = new RegExp(blazorCommentRegularExpression);
-    const definition = componentStartComment.exec(candidateStart.textContent);
+    const definition = blazorCommentRegularExpression.exec(candidateStart.textContent);
```
Well spotted - thanks for fixing this!

SteveSandersonMS commented on src/Components/Web.JS/src/webpack.config.js:
```diff
@@ -1,9 +1,14 @@
-    devtool: args.mode === 'development' ? 'source-map' : 'none',
+    resolve: { 
+        extensions: ['.ts', '.js'],
+
```
```suggestion
```

SteveSandersonMS commented on src/SignalR/clients/ts/signalr-protocol-msgpack/package.json:
```diff
@@ -44,9 +45,7 @@
   "dependencies": {
     "@microsoft/signalr": "link:../signalr",
-    "msgpack5": "^4.5.0"
+    "@msgpack/msgpack": "^2.4.0"
```
Just highlighting this to the SignalR team - not sure if you have any process to go through for the change of third-party dependency. @benaadams already updated the TPN file in this PR.

javiercn replied:
We talked about this on the last DoI. We are planning to have a process to upgrade these dependencies to their latest versions on a regular basis, except for maybe the typescript compiler in the SignalR repo since we want to make sure it works with older version of the typescript compiler.

BrennanConroy replied:
> We talked about this on the last DoI

Well we didn't talk about changing the third-party library we use for some of the implementation.

javiercn replied:
@BrennanConroy yeah, I was referring to updating the deps, changing the library you are right, that's up to you, but I think it's worth it.

benaadams replied:
> Well we didn't talk about changing the third-party library we use for some of the implementation.

It is moving to the "official" msgpack lib under the msgpack organisation; rather than going the otherway, if that helps? https://github.com/msgpack/msgpack-javascript

javiercn replied:
I think we are fine with this change, but I'll let @BrennanConroy confirm he's ok.

SteveSandersonMS commented:
Really superb, @benaadams - thanks a lot for contributing this!

benaadams committed:
Feedback

benaadams commented:
> Let me know when you are ready for me to review

@javiercn good to go

benaadams committed:
Hoist logger msg creation

halter73 commented on src/Components/Web.JS/src/Platform/Logging/Loggers.ts:
```diff
@@ -12,28 +12,29 @@ export class NullLogger implements Logger {
 
 export class ConsoleLogger implements Logger {
-  private readonly minimumLogLevel: LogLevel;
+  private readonly minLevel: LogLevel;
```
Why does this still show up in the terse output?

benaadams replied:
TypeScript private fields are just convention; they are still public on exported classes, so it doesn't drop the names.

![image](https://user-images.githubusercontent.com/1142958/108781280-8941cb00-7561-11eb-963a-4c57f389ab98.png)

It will mangle JavaScript private fields (ones which start with `#`); however if you use them in TypeScript it compiles them to references to a `WeakMap` rather than converting them to JavaScript private fields; which is unideal.

Terser does have a "Mangle properties" option; but that basically wrecks the output so it doesn't run if you don't play with the settings, was considering doing it in a follow-up when I workout what works.

benaadams replied:
Looks like might be able to solve it with
```js
mangle: {
    properties: {
        regex: /^_/
    }
},
```
and then prefixing every private field/method with `_`

benaadams committed:
SignalR: prefix private members with underscore and mangle

benaadams commented:
For SignalR prefix private members with underscore and mangling `_` prefixed properties shrinks it by another 6 kB

benaadams commented:
For SignalR prefix private members with underscore and mangling `_` prefixed properties shrinks it by another 6 kB

![image](https://user-images.githubusercontent.com/1142958/108805375-d9d41b00-7597-11eb-93e6-d74f3f2f7d88.png)

benaadams commented:
Can shave a further 11 kB off blazor.server that way; however its less happy and needs more work, so won't do it in this PR

javiercn reviewed (approved):
I think these changes are absolutely fantastic! I'm impressed beyond words on the outcome of this PR.

These are good to go on the Blazor side, as long as SignalR is happy with it, we are good to go here.

@benaadams thanks for all your hard work on this change, it's much appreciated!

benaadams committed:
Mark both .ts and .js as sideEffects

benaadams commented:
Added slight change for `sideEffects` using wildcard as the extension so it works for both the `.ts` and generated `.js` files

MaherJendoubi commented on src/SignalR/clients/ts/FunctionalTests/package.json:
```diff
@@ -30,9 +29,11 @@
+    "process": "^0.11.10",
     "rxjs": "^6.3.3",
     "ts-node": "^8.6.2",
     "typescript": "^3.7.5",
```
you could use the ^4.1.5 version for typescript

benaadams replied:
Done

MaherJendoubi replied:
Things are going faster. You could even use ^4.2.0 version of typescript which is just released.

MaherJendoubi commented on src/SignalR/clients/ts/signalr-protocol-msgpack/src/MessagePackOptions.ts:
```diff
@@ -3,21 +3,63 @@
+  /**
+   * @name forceFloat32 Force integers to be encoded as floats
+   */
+  forceIntegerToFloat?:	boolean;
```
There are some extra spaces here.

benaadams replied:
Done

MaherJendoubi commented on src/JSInterop/Microsoft.JSInterop.JS/src/package.json:
```diff
@@ -25,8 +26,8 @@
-    "tslint": "^5.9.1",
-    "typescript": "^2.7.1"
+    "rimraf": "^3.0.2",
+    "tslint": "^6.1.3",
```
TSLint has been deprecated in favor of ESLint.

benaadams replied:
Not changing linter; with different options, as part of this PR

MaherJendoubi commented on src/SignalR/clients/ts/FunctionalTests/package.json:
```diff
@@ -30,9 +29,11 @@
     "ts-node": "^8.6.2",
     "typescript": "^3.7.5",
+    "webpack": "^5.23.0",
     "ws": " ^6.0.0"
```
You could use the ^7.4.3 version for ws.

benaadams replied:
Done

BrennanConroy reviewed (commented):
```
warning " > tslint-consistent-codestyle@1.16.0" has incorrect peer dependency "tslint@^5.0.0".
warning "tslint-consistent-codestyle > @fimbul/bifrost@0.21.0" has incorrect peer dependency "tslint@^5.0.0".
```

Not sure why this doesn't fail the CI, but fails local build

BrennanConroy commented on src/SignalR/clients/ts/signalr-protocol-msgpack/tests/MessagePackHubProtocol.test.ts:
```diff
@@ -136,7 +136,7 @@ describe("MessagePackHubProtocol", () => {
             type: MessageType.StreamItem,
         } as StreamItemMessage],
-    ] as Array<[number[], StreamItemMessage]>).forEach(([payload, expectedMessage]) =>
+    ] as [number[], StreamItemMessage][]).forEach(([payload, expectedMessage]) =>
```
Line 231 of this file needs to be updated

BrennanConroy commented on src/SignalR/clients/ts/signalr/package.json:
```diff
@@ -46,17 +47,17 @@
+    "@types/node": "^14.14.31",
+    "@types/tough-cookie": "^4.0.0",
+    "process": "^0.11.10",
     "eventsource": "^1.0.7",
```
These need to be dependencies for Node apps to be able to use the library without bringing in polyfills manually.

benaadams replied:
Done

MaherJendoubi commented on src/Components/Web.JS/package.json:
```diff
@@ -14,22 +18,27 @@
     "test": "jest"
   },
   "devDependencies": {
+    "@babel/core": "^7.12.17",
```
updatable to ^7.13.1

benaadams replied:
Done

MaherJendoubi commented on src/Components/Web.JS/package.json:
```diff
@@ -14,22 +18,27 @@
   },
   "devDependencies": {
+    "@babel/core": "^7.12.17",
+    "@babel/preset-env": "^7.12.17",
```
updatable to ^7.13.5

benaadams replied:
Done

MaherJendoubi commented on src/Components/Web.JS/src/Platform/Circuits/DefaultReconnectionHandler.ts:
```diff
@@ -13,7 +14,7 @@ export class DefaultReconnectionHandler implements ReconnectionHandler {
     this._logger = logger;
     this._reconnectionDisplay = overrideDisplay;
-    this._reconnectCallback = reconnectCallback || (() => window['Blazor'].reconnect());
+    this._reconnectCallback = reconnectCallback || (() => (Blazor.reconnect as any)());
```
this._reconnectCallback = reconnectCallback || (() => (Blazor?.reconnect as any)());

benaadams replied:
Don't think that will change anything?

MaherJendoubi replied:
No, it doesn't.

MaherJendoubi commented on src/Components/Web.JS/babel.config.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "presets": [ [ "@babel/preset-env", { "targets": { "node": true } } ] ]
+}
```
Since [7.13.0 babel/core version](https://babeljs.io/blog/2021/02/22/7.13.0#top-level-targets-option-12189httpsgithubcombabelbabelpull12189-rfchttpsgithubcombabelrfcspull2), the config could be written this way:

`{
"targets": { "node": true },
"presets": ["@babel/preset-env"]  
}`

MaherJendoubi replied:
you could use the top-level targets option new feature for babel/core since version 7.13.0.

benaadams commented on src/SignalR/clients/ts/signalr/src/HubConnection.ts:
```diff
@@ -29,38 +29,39 @@ export enum HubConnectionState {
 export class HubConnection {
-    private readonly cachedPingMessage: string | ArrayBuffer;
+    private readonly _cachedPingMessage: string | ArrayBuffer;
+    // tslint:disable-next-line:naming-convention
```
The tests use the filed directly and adding an `_` will strip the name so is to allow it to not be stripped

BrennanConroy replied:
Ah yeah, our version of internals visible to I guess.

benaadams committed:
Feedback

benaadams commented on src/SignalR/clients/ts/common/package.json:
```diff
@@ -3,25 +3,29 @@
+    "typescript": "^4.1.5",
+    "terser": "^5.6.0",
+    "terser-webpack-plugin": "^5.1.1",
+    "webpack": "^5.23.0",
```
Done

MaherJendoubi replied:
Things are going faster. You could even use ^4.2.0 version of typescript which is just released.

benaadams replied:
Reverted the package.json changes to FunctionalTests as they broke the build and its not really a significant time of running them (which has to start various browser processes and .NET server apps as well as build them too)

benaadams commented on src/SignalR/clients/ts/FunctionalTests/package.json:
```diff
@@ -30,9 +29,11 @@
     "rxjs": "^6.3.3",
     "ts-node": "^8.6.2",
     "typescript": "^3.7.5",
+    "webpack": "^5.23.0",
```
Done

benaadams commented on src/Components/Web.JS/package.json:
```diff
@@ -14,22 +18,27 @@
+    "ts-jest": "^26.5.1",
+    "ts-loader": "^8.0.17",
+    "typescript": "^4.1.5",
+    "webpack": "^5.23.0",
```
Done

benaadams commented on src/Components/Web.JS/src/yarn.lock:
```diff
@@ -1,14 +1,6 @@
 # THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Done

benaadams commented on src/SignalR/clients/ts/FunctionalTests/junit.xml:
```diff
@@ -0,0 +1,295 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Done

benaadams commented on src/SignalR/clients/ts/signalr-protocol-msgpack/src/MessagePackHubProtocol.ts:
```diff
@@ -85,30 +85,28 @@ export class MessagePackHubProtocol implements IHubProtocol {
-        const msgpack = msgpack5(this.messagePackOptions);
-        // To avoid using the Buffer type we cast to 'any'. msgpack5 works with Uint8Array's
-        const properties = msgpack.decode(input as any);
+        const properties = decode(input, this._messagePackOptions) as any;
```
Done

benaadams commented on src/SignalR/clients/ts/signalr-protocol-msgpack/src/MessagePackOptions.ts:
```diff
@@ -3,21 +3,63 @@
    */
-  sortKeys?: boolean;
+  extensionCodec?: any;
+  /**
```
Done

benaadams commented on src/SignalR/clients/ts/signalr-protocol-msgpack/src/MessagePackOptions.ts:
```diff
@@ -3,21 +3,63 @@
+   */
+  forceFloat32?: boolean;
+  /**
+   * @name forceFloat32 Force integers to be encoded as floats
```
Done

benaadams commented on src/SignalR/clients/ts/signalr/src/Utils.ts:
```diff
@@ -11,7 +11,6 @@ import { Subject } from "./Subject";
 // Version token that will be replaced by the prepack command
 /** The version of the SignalR client. */
 export const VERSION: string = "0.0.0-DEV_BUILD";
-
```
Done

BrennanConroy commented on src/SignalR/clients/ts/tslint.json:
```diff
@@ -9,6 +10,10 @@
+        "no-console": false,
+        "naming-convention": [
+            true,
+            {"type": "member", "modifiers": "private", "leadingUnderscore": "require"}
```
Why was this removed?

benaadams replied:
Due to the build error you complained about

> warning " > tslint-consistent-codestyle@1.16.0" has incorrect peer dependency "tslint@^5.0.0".
warning "tslint-consistent-codestyle > @fimbul/bifrost@0.21.0" has incorrect peer dependency "tslint@^5.0.0".

TSLint is deprecated so didn't think it was worth solving and have already renamed all the variables

benaadams commented:
> Not sure why this doesn't fail the CI, but fails local build

Removed the checks; might want to add the `_` rule back as part of moving linter to ESLint

javiercn commented:
@BrennanConroy feel free to merge when you are happy.

BrennanConroy commented:
There are a couple broken npm unit tests.
You can run `yarn test` from the https://github.com/dotnet/aspnetcore/tree/main/src/SignalR/clients/ts directory to run them locally.

I have created PR https://github.com/dotnet/aspnetcore/pull/30409 to address the issue with them not showing up in the CI.

benaadams committed:
_transport -> transport (visible for tests)

benaadams commented:
> There are a couple broken npm unit tests.

Fixed, one was irrelevant and other was internals visible to (e.g. remove the `_` prefix)

benaadams committed:
Feedback

benaadams commented on src/Components/Web.JS/src/GlobalExports.ts:
```diff
@@ -1,10 +1,55 @@
+import { DefaultReconnectionHandler } from './Platform/Circuits/DefaultReconnectionHandler';
+import { CircuitStartOptions } from './Platform/Circuits/CircuitStartOptions';
+import { WebAssemblyStartOptions } from './Platform/WebAssemblyStartOptions';
+import { Platform } from '@microsoft/signalr/dist/esm/Utils';
```
Is used further down in the interface definition?

![image](https://user-images.githubusercontent.com/1142958/109073592-7f999e00-76ee-11eb-8e47-6551711fa7be.png)

BrennanConroy replied:
Right, but shouldn't it use the Platform type defined in Blazor?

benaadams replied:
Oooooh... 1 sec

benaadams replied:
Fixed

benaadams commented on src/Components/Web.JS/src/webpack.config.js:
```diff
@@ -14,5 +18,49 @@ module.exports = (env, args) => ({
     output: { path: path.join(__dirname, '/..', '/dist', args.mode == 'development' ? '/Debug' : '/Release'), filename: '[name].js' },
     performance: {
         maxAssetSize: 276000,
```
Reduced to `122880`

benaadams commented:
> Still here 😆

Removed this time 😀

benaadams commented:
Merged feedback commits to retrigger CI (transient error)

benaadams commented:
Same failures with submodules 😥
```
HEAD is now at 4ef7401480 Merge 1ba0197e771bb654b71dd5e705702fe98fa93914 into dd126d47b1a33fde1ce2c73c9f7d0ae2fddd9a7c
git submodule sync --recursive
git -c http.https://github.com.extraheader="AUTHORIZATION: basic ***" submodule update --init --force --recursive
Submodule 'src/submodules/MessagePack-CSharp' (https://github.com/aspnet/MessagePack-CSharp.git) registered for path 'src/submodules/MessagePack-CSharp'
Submodule 'googletest' (https://github.com/google/googletest) registered for path 'src/submodules/googletest'
Cloning into '/datadisks/disk1/workspace/_work/1/s/src/submodules/MessagePack-CSharp'...
fatal: unable to access 'https://github.com/aspnet/MessagePack-CSharp.git/'
  : Problem with the SSL CA cert (path? access rights?)
fatal: clone of 'https://github.com/aspnet/MessagePack-CSharp.git' into submodule path '/datadisks/disk1/workspace/_work/1/s/src/submodules/MessagePack-CSharp' failed
Failed to clone 'src/submodules/MessagePack-CSharp'. Retry scheduled
Cloning into '/datadisks/disk1/workspace/_work/1/s/src/submodules/googletest'...
fatal: unable to access 'https://github.com/google/googletest/'
  : Problem with the SSL CA cert (path? access rights?)
fatal: clone of 'https://github.com/google/googletest' into submodule path '/datadisks/disk1/workspace/_work/1/s/src/submodules/googletest' failed
Failed to clone 'src/submodules/googletest'. Retry scheduled
Cloning into '/datadisks/disk1/workspace/_work/1/s/src/submodules/MessagePack-CSharp'...
fatal: unable to access 'https://github.com/aspnet/MessagePack-CSharp.git/'
  : Problem with the SSL CA cert (path? access rights?)
fatal: clone of 'https://github.com/aspnet/MessagePack-CSharp.git' into submodule path '/datadisks/disk1/workspace/_work/1/s/src/submodules/MessagePack-CSharp' failed
Failed to clone 'src/submodules/MessagePack-CSharp' a second time, aborting
```

javiercn commented:
@pranavkm can you help out @benaadams here since you've been playing around this area?

benaadams commented:
I could try a rebase? Bit weird getting the submodules only fails on one platform though...

javiercn commented:
@benaadams I suspect it has to do with https://github.com/dotnet/aspnetcore/pull/30445

benaadams commented:
Sigh... now `apt-get` is failing
```
Get:17 http://archive.ubuntu.com/ubuntu bionic-backports/universe amd64 Packages [11.4 kB]
Get:18 http://archive.ubuntu.com/ubuntu bionic-backports/main amd64 Packages [11.3 kB]
Fetched 22.0 MB in 3s (6645 kB/s)
Reading package lists...
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python2.7/libpython2.7-minimal_2.7.17-1~18.04ubuntu1.3_amd64.deb  404  Not Found [IP: 91.189.88.142 80]
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python2.7/python2.7-minimal_2.7.17-1~18.04ubuntu1.3_amd64.deb  404  Not Found [IP: 91.189.88.142 80]
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python2.7/libpython2.7-stdlib_2.7.17-1~18.04ubuntu1.3_amd64.deb  404  Not Found [IP: 91.189.88.142 80]
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python2.7/python2.7_2.7.17-1~18.04ubuntu1.3_amd64.deb  404  Not Found [IP: 91.189.88.142 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
The command '/bin/sh -c apt-get update &&     apt-get -qqy install --no-install-recommends         jq         wget         locales         python         fakeroot         debhelper         build-essential         devscripts         unzip &&     rm -rf /var/lib/apt/lists/*' returned a non-zero code: 100

```

benaadams commented:
🥳

BrennanConroy commented:
Thanks @benaadams !
